File: mobile/src/main/java/org/openhab/habdroid/ui/ItemPickerAdapter.java
package org.openhab.habdroid.ui;

import android.content.Context;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.recyclerview.widget.RecyclerView;

import org.openhab.habdroid.R;
import org.openhab.habdroid.core.connection.Connection;
import org.openhab.habdroid.core.connection.ConnectionFactory;
import org.openhab.habdroid.core.connection.exception.ConnectionException;
import org.openhab.habdroid.model.Item;
import org.openhab.habdroid.ui.widget.WidgetImageView;
import org.openhab.habdroid.util.Constants;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;

public class ItemPickerAdapter extends RecyclerView.Adapter<ItemPickerAdapter.ItemViewHolder>
        implements View.OnClickListener {
    public interface ItemClickListener {
        void onItemClicked(Item item);
    }

    private ArrayList<Item> mFilteredItems = new ArrayList<>();
    private ArrayList<Item> mAllItems = new ArrayList<>();
    private ItemClickListener mItemClickListener;
    private final LayoutInflater mInflater;
    private int mHighlightedPosition = -1;
    private static String mIconFormat;

    public ItemPickerAdapter(Context context, ItemClickListener itemClickListener) {
        super();
        mIconFormat = PreferenceManager.getDefaultSharedPreferences(context)
                .getString(Constants.PREFERENCE_ICON_FORMAT, "");
        mInflater = LayoutInflater.from(context);
        mItemClickListener = itemClickListener;
    }

    public void setItems(List<Item> items) {
        mFilteredItems.clear();
        mFilteredItems.addAll(items);
        Collections.sort(mFilteredItems, new ItemNameComparator());
        mAllItems.clear();
        mAllItems.addAll(mFilteredItems);
        notifyDataSetChanged();
    }

    public void filter(String filter) {
        mFilteredItems.clear();
        String searchTerm = filter.toLowerCase();
        for (Item item : mAllItems) {
            if (item.getName().toLowerCase().contains(searchTerm)
                    || item.getLabel().toLowerCase().contains(searchTerm)
                    || item.getType().toString().toLowerCase().contains(searchTerm)) {
                mFilteredItems.add(item);
            }
        }
        notifyDataSetChanged();
    }

    private class ItemNameComparator implements Comparator<Item> {
        public int compare(Item left, Item right) {
            return left.getName().compareToIgnoreCase(right.getName());
        }
    }

    public void clear() {
        mFilteredItems.clear();
        notifyDataSetChanged();
    }

    public int findPositionForName(String name) {
        for (int i = 0; i < mFilteredItems.size(); i++) {
            if (TextUtils.equals(mFilteredItems.get(i).getName(), name)) {
                return i;
            }
        }
        return -1;
    }

    public void highlightItem(int position) {
        mHighlightedPosition = position;
        notifyItemChanged(position);
    }

    @NonNull
    @Override
    public ItemViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ItemViewHolder(mInflater, parent);
    }

    @Override
    public void onBindViewHolder(ItemViewHolder holder, int position) {
        holder.bind(mFilteredItems.get(position));
        holder.itemView.setOnClickListener(mItemClickListener != null ? this : null);

        if (position == mHighlightedPosition) {
            final View v = holder.itemView;
            v.post(() -> {
                if (v.getBackground() != null) {
                    final int centerX = v.getWidth() / 2;
                    final int centerY = v.getHeight() / 2;
                    DrawableCompat.setHotspot(v.getBackground(), centerX, centerY);
                }
                v.setPressed(true);
                v.setPressed(false);
                mHighlightedPosition = -1;
            });
        }
    }

    @Override
    public int getItemCount() {
        return mFilteredItems.size();
    }

    public static class ItemViewHolder extends RecyclerView.ViewHolder {
        final TextView mItemNameView;
        final TextView mItemLabelView;
        final WidgetImageView mIconView;
        final TextView mItemTypeView;

        public ItemViewHolder(LayoutInflater inflater, ViewGroup parent) {
            super(inflater.inflate(R.layout.itempickerlist_item, parent, false));
            mItemNameView = itemView.findViewById(R.id.itemName);
            mItemLabelView = itemView.findViewById(R.id.itemLabel);
            mItemTypeView = itemView.findViewById(R.id.itemType);
            mIconView = itemView.findViewById(R.id.itemIcon);
            itemView.setTag(this);
        }

        public void bind(Item item) {
            mItemNameView.setText(item.getName());
            mItemLabelView.setText(item.getLabel());
            mItemTypeView.setText(item.getType().toString());

            Connection connection;
            try {
                connection = ConnectionFactory.Companion.getUsableConnection();
            } catch (ConnectionException e) {
                connection = null;
            }
            if (item.getCategory() != null && connection != null) {
                String iconUrl = String.format(Locale.US, "images/%s.%s",
                        Uri.encode(item.getCategory()),
                        TextUtils.isEmpty(mIconFormat) ? "png" : mIconFormat);
                mIconView.setImageUrl(connection, iconUrl, mIconView.getResources()
                        .getDimensionPixelSize(R.dimen.notificationlist_icon_size), 2000);
            } else {
                mIconView.setImageResource(R.drawable.ic_openhab_appicon_24dp);
            }
        }
    }

    @Override
    public void onClick(View view) {
        ItemPickerAdapter.ItemViewHolder holder = (ItemPickerAdapter.ItemViewHolder) view.getTag();
        int position = holder.getAdapterPosition();
        if (position != RecyclerView.NO_POSITION) {
            mItemClickListener.onItemClicked(mFilteredItems.get(position));
        }
    }
}

File: mobile/src/main/java/org/openhab/habdroid/ui/ItemPickerAdapter.kt
/*
 * Copyright (c) 2010-2020 Contributors to the openHAB project
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 */

package org.openhab.habdroid.ui

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import org.openhab.habdroid.R
import org.openhab.habdroid.core.connection.ConnectionFactory
import org.openhab.habdroid.model.Item
import org.openhab.habdroid.model.toOH2IconResource
import org.openhab.habdroid.ui.widget.WidgetImageView
import org.openhab.habdroid.util.isDataSaverActive
import java.util.ArrayList
import java.util.Comparator
import java.util.Locale

class ItemPickerAdapter(context: Context, private val itemClickListener: ItemClickListener?) :
    RecyclerView.Adapter<ItemPickerAdapter.ItemViewHolder>(), View.OnClickListener {

    private val filteredItems = ArrayList<Item>()
    private val allItems = ArrayList<Item>()
    private val inflater = LayoutInflater.from(context)
    private var highlightedPosition = -1

    interface ItemClickListener {
        fun onItemClicked(item: Item)
    }

    fun setItems(items: List<Item>) {
        filteredItems.clear()
        filteredItems.addAll(items)
        filteredItems.sortWith(Comparator { lhs, rhs -> lhs.name.compareTo(rhs.name, ignoreCase = true) })
        allItems.clear()
        allItems.addAll(filteredItems)
        notifyDataSetChanged()
    }

    fun filter(filter: String) {
        filteredItems.clear()
        val searchTerm = filter.toLowerCase(Locale.getDefault())
        allItems.filterTo(filteredItems) { item ->
            searchTerm in item.name.toLowerCase(Locale.getDefault()) ||
                searchTerm in item.label?.toLowerCase(Locale.getDefault()).orEmpty() ||
                searchTerm in item.type.toString().toLowerCase(Locale.getDefault())
        }
        notifyDataSetChanged()
    }

    fun clear() {
        filteredItems.clear()
        notifyDataSetChanged()
    }

    fun findPositionForName(name: String): Int {
        return filteredItems.indexOfFirst { item -> item.name == name }
    }

    fun highlightItem(position: Int) {
        highlightedPosition = position
        notifyItemChanged(position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
        return ItemViewHolder(inflater, parent)
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        holder.bind(filteredItems[position])
        holder.itemView.setOnClickListener(if (itemClickListener != null) this else null)

        if (position == highlightedPosition) {
            holder.itemView.playPressAnimationAndCallBack {
                highlightedPosition = -1
            }
        }
    }

    override fun getItemCount(): Int {
        return filteredItems.size
    }

    override fun onClick(view: View) {
        val holder = view.tag as ItemViewHolder
        val position = holder.adapterPosition
        if (position != RecyclerView.NO_POSITION) {
            itemClickListener?.onItemClicked(filteredItems[position])
        }
    }

    class ItemViewHolder(inflater: LayoutInflater, parent: ViewGroup) :
        RecyclerView.ViewHolder(inflater.inflate(R.layout.itempickerlist_item, parent, false)) {
        private val itemNameView: TextView = itemView.findViewById(R.id.itemName)
        private val itemLabelView: TextView = itemView.findViewById(R.id.itemLabel)
        private val iconView: WidgetImageView = itemView.findViewById(R.id.itemIcon)
        private val itemTypeView: TextView = itemView.findViewById(R.id.itemType)

        init {
            itemView.tag = this
        }

        fun bind(item: Item) {
            itemNameView.text = item.name
            itemLabelView.text = item.label
            itemTypeView.text = item.type.toString()

            val connection = ConnectionFactory.usableConnectionOrNull
            val icon = item.category.toOH2IconResource()
            if (icon != null && connection != null) {
                val size = iconView.resources.getDimensionPixelSize(R.dimen.notificationlist_icon_size)
                iconView.setImageUrl(
                    connection,
                    icon.toUrl(itemView.context, !itemView.context.isDataSaverActive()),
                    size,
                    2000
                )
            } else {
                iconView.setImageResource(R.drawable.ic_openhab_appicon_24dp)
            }
        }
    }
}
